DUI3-DX/Converters/Rhino/Speckle.Converters.Rhino7/Geometry/BrepToSpeckleConverter.cs

```diff
@@ -0,0 +1,200 @@
+    _intervalConverter = intervalConverter;
+    _contextStack = contextStack;
+  }
+
+  public Base Convert(object target) => RawConvert((RG.Brep)target);
+
+  public SOG.Brep RawConvert(RG.Brep target)
+  {
+    var tol = _contextStack.Current.Document.ModelAbsoluteTolerance;
+    target.Repair(tol);
+
+    // POC: This should come as part of the user settings in the context object.
+    // if (PreprocessGeometry)
+    // {
+    //   brep = BrepEncoder.ToRawBrep(brep, 1.0, Doc.ModelAngleToleranceRadians, Doc.ModelRelativeTolerance);
+    // }
+
+    // get display mesh and attach render material to it if it exists
+    var displayMesh = GetBrepDisplayMesh(target);
+    var displayValue = new List<SOG.Mesh>();
+    if (displayMesh != null)
+    {
+      displayValue.Add(_meshConverter.RawConvert(displayMesh));
+    }
+
+    // POC: Swap input material for something coming from the context.
+    // if (displayValue != null && mat != null)
+    // {
+    //   displayValue["renderMaterial"] = mat;
+    // }
+
+    // Vertices, uv curves, 3d curves and surfaces
+    var vertices = target.Vertices.Select(vertex => _pointConverter.RawConvert(vertex.Location)).ToList();
+    var curves3d = target.Curves3D.Select(curve3d => _curveConverter.RawConvert(curve3d)).ToList();
+    var surfaces = target.Surfaces.Select(srf => _surfaceConverter.RawConvert(srf.ToNurbsSurface())).ToList();
+
+    List<ICurve> curves2d;
+    using (_contextStack.Push(Units.None))
+    {
+      // Curves2D are unitless, so we convert them within a new pushed context with None units.
+      curves2d = target.Curves2D.Select(curve2d => _curveConverter.RawConvert(curve2d)).ToList();
+    }
+
+    var speckleBrep = new SOG.Brep
+    {
+      Vertices = vertices,
+      Curve3D = curves3d,
+      Curve2D = curves2d,
+      Surfaces = surfaces,
+      displayValue = displayValue,
+      IsClosed = target.IsSolid,
+      Orientation = (SOG.BrepOrientation)target.SolidOrientation,
+      volume = target.IsSolid ? target.GetVolume() : 0,
+      area = target.GetArea(),
+      bbox = _boxConverter.RawConvert(new RG.Box(target.GetBoundingBox(false))),
+      units = _contextStack.Current.SpeckleUnits
+    };
+
+    // Brep non-geometry types
+    var faces = ConvertBrepFaces(target, speckleBrep);
+    var edges = ConvertBrepEdges(target, speckleBrep);
+    var loops = ConvertBrepLoops(target, speckleBrep);
+    var trims = ConvertBrepTrims(target, speckleBrep);
+
+    return speckleBrep;
+  }
+
+  private static List<SOG.BrepFace> ConvertBrepFaces(RG.Brep brep, SOG.Brep speckleParent) =>
+    brep.Faces
+      .Select(
+        f =>
+          new SOG.BrepFace(
+            speckleParent,
+            f.SurfaceIndex,
+            f.Loops.Select(l => l.LoopIndex).ToList(),
+            f.OuterLoop.LoopIndex,
+            f.OrientationIsReversed
+          )
+      )
+      .ToList();
+
+  private List<SOG.BrepEdge> ConvertBrepEdges(RG.Brep brep, SOG.Brep speckleParent) =>
```

Comment:
I wonder with these elements whether these should be injected. We could be testing these in isolation? I guess we might have tests already we are breaking?

Reply:
Opened a ticket to look at refactoring BREP conversion separately https://spockle.atlassian.net/browse/CNX-9160